Automattic/jetpack
# Fix Jetpack carousel data enrichment

Fixes Automattic/wp-calypso#64638
Fixes #22726

#### Changes proposed in this Pull Request:

* This PR changes the way we add additional data attributes for carousel-enabled gallery blocks so that we can correctly add the data for reusable blocks.
* The existing code was using `'the_content'` filter to add the `data-carousel-extra` attribute to the outermost gallery element, but this meant that it was not being run for reusable blocks, because the main post content won't contain the resolved block content yet
* The new code uses the [`render_block_{name}`](https://developer.wordpress.org/reference/hooks/render_block_this-name/) filter to add the data attributes to the first HTML element for the gallery - this now works for reusable and standard blocks

##### Possible compatibility issues

1. We now add the data attributes to the first HTML element, which is different to the previous code, which might add it to up to 4 elements in the gallery tree. Nobody _should_ have been relying on this behaviour, but it was still there. Furthermore, we now perform this addition for each block independently instead of once for all the content.
2. We call the `jp_carousel_add_data_to_container` filter for every gallery block, instead of once for all the content. This may be unexpected for some use cases.
3. We've removed the `check_content_for_blocks()` function from the `Jetpack_Carousel` class. This was marked as public, so may have been used by someone.

#### Other information:


#### Jetpack product discussion

This is a bugfix rather than a product change.

#### Does this pull request change what data or activity we track or use?

The PR doesn't affect what data we record or activity we track.

#### Testing instructions:

* Create a test site, and ensure that it has the Jetpack and Jetpack Beta plugins installed
* Ensure that the Carousel feature is enabled via the Media section of the _Writing_ tab in Jetpack Settings
* Set up the test site so that you have four separate posts with each post containing exactly one of the following blocks:
  - A standard Gallery block with at least two images
  - A standard Tiled Gallery block with at least two images
  - A reusable block that contains a Gallery block with at least two images
  - A reusable block that contains a Tiled Gallery block with at least two images
* Open each of the four posts above in new tabs
* Verify that the two base galleries open the carousel when an image in the gallery is clicked on, and the two galleries within reusable blocks do _not_ open the carousel when any images are clicked on
  - You can also take a look at the markup to see the cases where we have a profusion of `data-carousel-extra` attributes in the element tree
* Use the Jetpack Beta tool to switch to the code for this branch (`update/jetpack-gallery-data-enrichment`)
* Refresh each of the four pages, and verify that all four galleries open the carousel when you click on an image
* Inspect each of the gallery elements, and verify that only one `data-carousel-extra` attributes exists for each gallery element

##### Regression tests

We also want to make sure we're not introducing any regressions, as noted in [this comment](https://github.com/Automattic/jetpack/pull/25441#issuecomment-1214684123). To that end, verify that all of the following cases continue to work as expected:
* A post with a Classic block that contains a standard gallery
  - After adding a Classic block, you can create a standard gallery by clicking on "Add Media" in the toolbar, selecting multiple images, and then selecting "Thumbnail Gallery" in the _Layout_ dropdown
* A post with a Classic block that contains a tiled gallery
  - After adding a Classic block, you can create a tiled gallery by clicking on "Add Media" in the toolbar, selecting multiple images, and then specifying a "Tiled Mosaic" in the _Layout_ dropdown
* A post with a Classic block that contains single image
  - After adding a Classic block, click on "Add Media" and select a single image
* A post with an Image block with the "Lightbox" toggle enabled
* A post containing all four of the above blocks as well as a gallery block and a tiled gallery block
 - This should ensure we don't get interference between the block and shortcode processing

In all of the cases above, except for the single image inside the classic block, clicking on the image should show a carousel (or a lightbox for the single image block).
